### PR TITLE
SEC-78 Restrict symbol table interface

### PR DESCRIPTION
**Pull request by smowton**
This restricts the interface of symbol_tablet to prevent direct write operations (instead routing them via new public functions), which makes it easier to introduce variants of symbol_tablet with different behaviour; for example, the journalling implementation that is forthcoming in a subsequent PR.

This upstreams https://github.com/diffblue/security-scanner/pull/205 -- it was already reviewed by @thk123 and @reuk over there.

@NathanJPhillips recommend reviewing to check I haven't fouled anything up, @thk123 might want to look again to check this is acceptable for upstream; tagging @mgudemann as a fresh reviewer just for the upstreaming.

### CODE REVIEW

**Review by NathanJPhillips — APPROVED** [review]

**Comment by smowton** [comment]
@thk123 I'm making a test-gen corresponding PR now

**Comment by smowton** [comment]
Companion test-gen PR created at https://github.com/diffblue/test-gen/pull/1086

File: src/cpp/cpp_typecheck_compound_type.cpp [hunk 1]
```diff
@@ -540,9 +536,8 @@ void cpp_typecheckt::typecheck_compound_declarator(
         vt_symb_type.type.set(ID_name, vt_symb_type.name);
         vt_symb_type.is_type=true;
 
-        bool failed=symbol_table.move(vt_symb_type);
+        const bool failed=!symbol_table.insert(std::move(vt_symb_type)).second;
         assert(!failed);
```
Comment (mgudemann) [inline src/cpp/cpp_typecheck_compound_type.cpp:540]: make this an invariant?

File: src/cpp/cpp_typecheck_compound_type.cpp [hunk 2]
```diff
@@ -676,7 +670,7 @@ void cpp_typecheckt::typecheck_compound_declarator(
 
         // add the function to the symbol table
         {
-          bool failed=symbol_table.move(func_symb);
+          bool failed=!symbol_table.insert(std::move(func_symb)).second;
```
Comment (mgudemann) [inline src/cpp/cpp_typecheck_compound_type.cpp]: `const bool` here, too?

File: src/goto-instrument/goto_program2code.cpp [hunk 3]
```diff
@@ -1671,12 +1668,12 @@ void goto_program2codet::remove_const(typet &type)
     if(!const_removed.insert(identifier).second)
       return;
 
-    symbol_tablet::symbolst::iterator it=
-      symbol_table.symbols.find(identifier);
-    assert(it!=symbol_table.symbols.end());
-    assert(it->second.is_type);
+    symbolt &symbol=*symbol_table.get_writeable(identifier);
+    INVARIANT(
```
Comment (mgudemann) [inline src/goto-instrument/goto_program2code.cpp:1672]: the second assert (`..!=..end()`)has vanished, is this an invariant in `get_writeable` ?

File: src/pointer-analysis/add_failed_symbols.cpp [hunk 4]
```diff
@@ -45,28 +39,32 @@ void add_failed_symbol(symbolt &symbol, symbol_tablet &symbol_table)
     if(new_symbol.type.id()==ID_pointer)
       add_failed_symbol(new_symbol, symbol_table); // recursive call
 
-    symbol_table.move(new_symbol);
+    symbol_table.insert(std::move(new_symbol));
   }
 }
 
+void add_failed_symbol(const symbolt &symbol, symbol_tablet &symbol_table)
```
Comment (mgudemann) [inline src/pointer-analysis/add_failed_symbols.cpp:46]: add documentation

File: src/util/symbol_table.cpp [hunk 5]
```diff
@@ -1,30 +1,48 @@
...
-
-  symbol_base_map.insert(
-    std::pair<irep_idt, irep_idt>(symbol.base_name, symbol.name));
-  symbol_module_map.insert(
-    std::pair<irep_idt, irep_idt>(symbol.module, symbol.name));
+  return !insert(symbol).second;
+}
 
-  return false;
+std::pair<symbolt &, bool> symbol_tablet::insert(symbolt symbol)
```
Comment (mgudemann) [inline src/util/symbol_table.cpp]: document function

File: src/util/symbol_table.cpp [hunk 6]
```diff
@@ -44,65 +62,71 @@ bool symbol_tablet::add(const symbolt &symbol)
...
-      break;
-    }
-
-  symbols.erase(entry);
-
+  erase(entry);
   return false;
 }
 
+void symbol_tablet::erase(const symbolst::const_iterator &entry)
```
Comment (mgudemann) [inline src/util/symbol_table.cpp]: document

**Review by mgudemann — APPROVED** [review]
I have some very minor comments, nothing that prevents merging. Documentation for new functions should be added, I've later seen that some are already documented in the respective header.

File: src/goto-instrument/goto_program2code.cpp (hunk 3 above)
Comment (smowton) [reply src/goto-instrument/goto_program2code.cpp:1672]: No, but it is of `operator*` against an `optionalt`

**Comment by smowton** [comment]
I'll leave invariants for another day, but other comments applied

**Comment by smowton** [comment]
Merged as @thk123 had reviewed its predecessor already and the test-gen bump addressed his concern that this may break test-gen.

**Pull request merged**